lib/AST/NestedCondProp.cpp

```diff
@@ -12,106 +12,208 @@
 #include <gflags/gflags.h>
 #include <glog/logging.h>
 
+#include <unordered_map>
 #include <vector>
 
 #include "rellic/AST/ASTBuilder.h"
 #include "rellic/AST/Util.h"
 
 namespace rellic {
-using ExprVec = std::vector<clang::Expr*>;
+// Stores a set of expression that have a known value, so that they can be
+// recognized as part of larger expressions and simplified.
+struct KnownExprs {
+  std::unordered_map<unsigned, bool> values;
```

Comment:
What about using the hash specialisation trick here too? Or better yet, make an `ExprMap` and put it in `AST/Utils.cpp`.

Reply:
The issue with using the hash specialization trick is that `unordered_map` still needs to store the actual `z3::expr` objects in order to call the specializations on them, and at that point we're back to having to use an `expr_vector` I think

Reply:
Ah, fair.